docs/reference/style-guides/node-style-guide.md

```diff
@@ -2,8 +2,186 @@
 title: Node Style Guide
 ---
 
-Current as of `July 18, 2022`
+Current as of `January 18, 2023`
 
-## Introduction
+## Code Standards
 
-Building...
+These code standards apply to new packages in the FxA monorepo, such as:
+
+- `packages/fxa-graphql-api`
+- `packages/fxa-event-broker`
+
+Most of the standards in the code is enforced with [Prettier][2] and [ESLint][1].
+
+### Style
+
+We use the [Google TypeScript Style Guide](https://google.github.io/styleguide/tsguide.html#source-organization) for our base code style.
+
+Code should also follow the [Do's and Don'ts of TypeScript](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html).
+
+### Structure
+
+#### Organize by Feature
+
+(This is present in the Google TypeScript Style Guide, but bears repeating)
+
+Code should be organized by feature, not by type.
+
+Don't:
+
+```
+app/
+├── controllers/
+│   └── auth.ts
+├── models/
+│   └── auth.ts
+├── validators/
+│   └── auth.ts
+```
+
+Do:
+
+```
+app/
+├── auth/
+│   ├── auth.controller.ts
+│   ├── auth.model.ts
+│   └── auth.validator.ts
+```
+
+If a feature has many DTOs, models, etc., a hybrid approach may be used, organized by feature first:
+
+```
+app/
+├── auth/
+│   ├── dto/
+│   ├── model/
+│   |   └── auth.model.ts
+│   ├── auth.controller.ts
+│   └── auth.validator.ts
+```
+
+Tests should be organized alongside the code being tested, with the `.spec.ts` suffix:
+
+```
+app/
+├── auth/
+│   ├── auth.controller.ts
+│   ├── auth.controller.spec.ts
+```
+
+#### NestJS File Naming Conventions
+
+##### File Naming for Class
+
+```typescript
+export class PascalCaseSuffix {} //= pascal-case.suffix.ts
+// Except for suffix, PascalCase to hyphen-case
+class FooBarNaming {} //= foo-bar.naming.ts
+class FooController {} //= foo.controller.ts
+class BarQueryDto {} //= bar-query.dto.ts
+```
+
+#### Interface Naming
+
+```typescript
+// https://stackoverflow.com/questions/541912
+// https://stackoverflow.com/questions/2814805
+interface User {}
+interface CustomUser extends User {}
+interface ThirdCustomUser extends CustomUser {}
+```
+
+#### Index Exporting
+
+```diff
+# It is recommended to place index.ts in each folder and export.
+# Unless it's a special case, it is imported from a folder instead of directly from a file.
+- import { FooController } from './controllers/foo.controller';
+- import { BarController } from './controllers/bar.controller';
++ import { FooController, BarController } from './controllers';
+# Preferred method is to place only one file or folder name at the end of the path.
+- import { UtilService } from '../common/providers/util.service';
++ import { UtilService } from '../common';
+```
+
+### TypeScript
+
+#### Tests
+
+All tests should be written in TypeScript.
+
+#### Never use Type Asserts
+
+Type asserts should not be used in any situation. Instead, use one of the following options:
```

Comment:
In addition to showing people what to do, I think an example of what not to do would be merited.

Reply:
```typescript
function example(arg?: string) {
  arg!.substring(0, 20);
}
```

Reply:
Ah, possibly two examples of don'ts then, as I was thinking more:
```typescript
function example(): MyType {
  # Perhaps otherFunc returns a union type and we have good reason to believe
  # it will always be MyType
  return otherFunc() as MyType;
}
```

Reply:
Ah, we may want to break this into two separate sections then - one covering type narrowing for nulls (where the "don't" is non-nullish assertions), and another for type narrowing multiple types (or generic types) into a single type. The "don't" example you gave @bbangert would probably need to be addressed as such:

```
interface MyType {
  prop: string
}
interface AnotherType {
  prop2: string
}
function example(arg: MyType | AnotherType) {
  if ('prop' in arg) {
    // arg is now of type MyType
  } else {
    // arg is now of type AnotherType
  }
}
```

Reply:
Ok, added examples of these.